### PR TITLE
font: add anti-aliasing and other features

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

* Allow fonts to be multi-bit per pixel for anti-aliasing purposes;
* Allow additional glyphs from other font files to be merged into a font;
* Allow additional font file types;
* Improve font encoding, reduces size of binary encoding.

These changes are usable now in the existing `display` component using lambdas, but also support the upcoming LVGL component which can make use of fonts with more than one bit per pixel.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3596

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
font:
  - file: "gfonts://Roboto"
    id: roboto
    size: 20
    glyphs: "0123456789."
    extras:
      - file: "gfonts://Roboto"
        glyphs: [
          "\u00C4",
          "\u00C5",
          "\U000000C7"
          ]

display:
  - platform: ...
    id: displ8
    lambda: |-
      it.printf(10, 100, id(roboto), Color(0x123456), COLOR_OFF, display::TextAlign::BASELINE, "%f", id(heap_free).state);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
Tested and working:
```yaml
    pages:
    - id: graph_page
      lambda: |-
        it.fill(Color(255,255,255));
        it.print(it.get_width()/2, it.get_height()/4 - 25, id(roboto_icons_56_1bpp), display::COLOR_OFF, TextAlign::CENTER, "Just\U000f05d4here. Already\U000F02D1this.");
        it.print(it.get_width()/2, it.get_height()/4 + 25, id(roboto_icons_56_4bpp), display::COLOR_OFF, TextAlign::CENTER, "Just\U000f05d4here. Already\U000F02D1this.", Color(255,255,255));

font:
  - file: 'fonts/RobotoCondensed-Regular.ttf'
    id: roboto_icons_56_1bpp
    size: 56
    bpp: 1
    extras:
      - file: 'fonts/materialdesignicons-webfont.ttf'
        glyphs: [
          "\U000F02D1", # mdi-heart
          "\U000F05D4", # mdi-airplane-landing
          ]

  - file: 'fonts/RobotoCondensed-Regular.ttf'
    id: roboto_icons_56_4bpp
    size: 56
    bpp: 4
    extras:
      - file: 'fonts/materialdesignicons-webfont.ttf'
        glyphs: [
          "\U000F02D1", # mdi-heart
          "\U000F05D4", # mdi-airplane-landing
          ]
```
results in:
kép

**Comment by mathieu-mp** [comment]
> kép

Hi @nagyrobi,
Could you give me a lead on how to generate this kind of output for the display component (generated computer-side and not on the actual hardware), as I am working on esphome/esphome#6108 by building and uploading to the ESP32 for each trial and error?
I know this is so newbie from me that I allowed myself to ask on this PR, sorry for that! (delete this comment if I am bothering here).

**Comment by nagyrobi** [comment]
Hi, you can use the host platform to generate a linux executable, and the `vnc` display and touchscreen platforms to connect to it remotely. Looks like it's in beta too.
See: https://github.com/esphome/esphome/pull/6137
(no need to use lvgl, vnc shows the classic display renderer too)

File: esphome/components/font/__init__.py [hunk 1]
```diff
@@ -79,16 +112,16 @@ def validate_pillow_installed(value):
     return value
 
 
+FONT_EXTENSIONS = (".ttf", ".woff", ".otf")
```
Comment (nagyrobi) [inline esphome/components/font/__init__.py:115]: FONT_EXTENSIONS = (".ttf", ".woff", ".otf", ".TTF", ".WOFF", ".OTF")

Comment (nagyrobi) [reply esphome/components/font/__init__.py:115]: For users in the Windows world

Comment (nagyrobi) [reply esphome/components/font/__init__.py:115]: Or, change case to small in the code, before compare

**Review by kbx81 — APPROVED** [review]

**Pull request merged**